### ISSUE TITLE
An urgent piece of advice ...

### ISSUE THREAD

**Issue by mythsman** [mde/ejs#357]
Well ,  after publishing a backward incompatible version , i think you'd better improve your major version number like '3.0.0' instead of '2.x.x' in npm.

I have been using your package for some time and i think it really convenient, but today when i publish my code , i find it just don't work . After debugging for quite a while , i find the bug is that you change the usage of the function 'ejs.renderFile' which is not compatible with the old version ...

Actually , in npm  we usually use the version number like '^2.5.6' which will automatically improve to '2.5.7' or '2.x.x' but never to '3.0.0' .So if you publish a backward incompatible version ,I suggest you improve the major version number....:smile:

**Comment by mde** [comment]
Could you be a tiny bit more specific about what is broken?

Semver (https://semver.org/), which is what you're talking about, very clearly refers to changes to the public or documented API, and there was no intentional change here to documented behavior of any public API.

And of course, even the most rigorous following of semver does not guarantee there will be no bugs when making changes.

I'm happy to push a fix if you'll give me some sort of idea of what problem you're seeing.

**Comment by mythsman** [comment]
Well , maybe it blames to my misuse of the package.
In version '2.5.7' and below ,the following code works, and will output the correct result.
```javascript
var ejs = require('ejs')

var contents = ejs.renderFile('index.ejs', {
    value: 'hello world'
}, function (err, result){
    if (err) console.log(err);
    return result;
});

console.log(contents);
```
But in version '2.5.8', it will output 'undifined' . 
This is the problem i've met.
Thank you for your help~~~

**Comment by mde** [comment]
Ah, so this is definitely undocumented functionality. Also, a callback-based API is usually what you use when things happen asynchronously, and you can't get a value directly returned from a function call. (The callback is the thing that receives the value and acts on it, since the value can't be returned.)

In your example, you would normally do the console.log of `contents` directly inside your callback, rather than returning a value and then using it later.

This is really the entire point of using a callback-based API.

However, having said that, you're not alone in this. Your issue here is a duplicate of #355. Other folks have tripped over this change. I'm not super excited about the idea of restoring the old behavior, because it encourages use of an async API in a synchronous way, which will make it harder to make it truly async in the future.

OTOH, that sort of change would almost certainly be a major version bump, so maybe it's okay to allow people to rely on the current behavior. In any case, I'll close this issue, and we can continue any discussion over on issue #355.

**Comment by mde** [comment]
Closed as a duplicate of #355.
